guava/src/com/google/common/hash/BloomFilterStrategies.java

```diff
@@ -138,80 +140,165 @@
       }
       return false;
     }
 
-    boolean get(long index) {
-      return (data[(int) (index >>> 6)] & (1L << index)) != 0;
+    boolean get(long bitIndex) {
+      return (data.get((int) (bitIndex >>> 6)) & (1L << bitIndex)) != 0;
+    }
+
+    /**
+     * Careful here: if threads are mutating the atomicLongArray while this
+     * method is executing, the final long[] will be "rolling snapshot" of the
+     * state of the bit array. This is usually good enough, but should be kept in mind.
+     */
+    public static long[] toPlainArray(AtomicLongArray atomicLongArray) {
+      long[] array = new long[atomicLongArray.length()];
+      for (int i = 0; i < array.length; ++i) {
+        array[i] = atomicLongArray.get(i);
+      }
+      return array;
     }
 
     /** Number of bits */
     long bitSize() {
-      return (long) data.length * Long.SIZE;
+      return (long) data.length() * Long.SIZE;
     }
 
-    /** Number of set bits (1s) */
+    /**
+     * Number of set bits (1s).
+     *
+     * Note that because of concurrent set calls and uses of atomics, this
+     * bitCount is a (very) close *estimate* of the actual number of bits set.
+     * It's not possible to do better than an estimate without locking. Note
+     * that the number, if not exactly accurate, is *always*
+     * underestimating, never overestimating.
+     */
     long bitCount() {
-      return bitCount;
+      return bitCount.longValue();
     }
 
-    BitArray copy() {
-      return new BitArray(data.clone());
+    LockFreeBitArray copy() {
+      return new LockFreeBitArray(toPlainArray(data));
     }
 
-    /** Combines the two BitArrays using bitwise OR. */
-    void putAll(BitArray array) {
+    /**
+     * Combines the two BitArrays using bitwise OR.
+     *
+     * NOTE: Because of the use of atomics, if the other LockFreeBitArray is being
+     * mutated while this operation is executing, not all of those new 1's may be
+     * set in the final state of this LockFreeBitArray. The ONLY guarantee
+     * provided is that all the bits that were set in the other LockFreeBitArray at
+     * the start of this method will be set in this LockFreeBitArray at the end of this method.
+     */
+    void putAll(LockFreeBitArray other) {
       checkArgument(
-          data.length == array.data.length,
+          data.length() == other.data.length(),
           "BitArrays must be of equal length (%s != %s)",
-          data.length,
-          array.data.length);
-      bitCount = 0;
-      for (int i = 0; i < data.length; i++) {
-        data[i] |= array.data[i];
-        bitCount += Long.bitCount(data[i]);
+          data.length(),
+          other.data.length());
+      for (int i = 0; i < data.length(); i++) {
+        long otherLong = other.data.get(i);
+
+        long ourLongOld, ourLongNew;
+        do {
+          ourLongOld = data.get(i);
+          ourLongNew = ourLongOld | otherLong;
+          if (ourLongOld == ourLongNew) {
+            continue;
```

Comment:
What I wanted was to continue from the outer loop, but somehow forgot I was in the inner one. We still perform the bitcount ops unnecessarily though. I could use a goto to avoid this, but I'd rather not.